art/estimators/speech_recognition/pytorch_espresso.py

```diff
@@ -0,0 +1,389 @@
+from art.estimators.pytorch import PyTorchEstimator
+from art.estimators.speech_recognition.speech_recognizer import SpeechRecognizerMixin
+
+from pathlib import Path
+
+if TYPE_CHECKING:
+    import torch
+    
+    from art.defences.preprocessor.preprocessor import Preprocessor
+    from art.defences.postprocessor.postprocessor import Postprocessor
+    from art.utils import CLIP_VALUES_TYPE, PREPROCESSING_TYPE
+
+logger = logging.getLogger(__name__)
+
+INT16MAX = 32767
+
+
+class PyTorchEspresso(SpeechRecognizerMixin, PyTorchEstimator):
+    """
+    This class implements a model-specific automatic speech recognizer using the end-to-end speech recognizer
+    in Espresso.
+
+    | Paper link: https://arxiv.org/abs/1909.08723
+    """
+
+    estimator_params = PyTorchEstimator.estimator_params
+
+    def __init__(
+        self,
+        espresso_config_filepath: str,
+        clip_values: Optional["CLIP_VALUES_TYPE"] = None,
+        preprocessing_defences: Union["Preprocessor", List["Preprocessor"], None] = None,
+        postprocessing_defences: Union["Postprocessor", List["Postprocessor"], None] = None,
+        preprocessing: "PREPROCESSING_TYPE" = None,
+        device_type: str = "gpu",
+    ):
+        """
+        Initialization of an instance PyTorchEspresso
+        :param espresso_config_filepath: The path of the espresso config file (yaml) # TODO: the config file is not automatically generated in Espresso. It was manually created by us.
+        :param clip_values: Tuple of the form `(min, max)` of floats or `np.ndarray` representing the minimum and
+               maximum values allowed for features. If floats are provided, these will be used as the range of all
+               features. If arrays are provided, each value will be considered the bound for a feature, thus
+               the shape of clip values needs to match the total number of features.
+        :param preprocessing_defences: Preprocessing defence(s) to be applied by the estimator.
+        :param postprocessing_defences: Postprocessing defence(s) to be applied by the estimator.
+        :param preprocessing: Tuple of the form `(subtrahend, divisor)` of floats or `np.ndarray` of values to be
+               used for data preprocessing. The first value will be subtracted from the input. The input will then
+               be divided by the second one.
+        :param device_type: Type of device to be used for model and tensors, if `cpu` run on CPU, if `gpu` run on GPU
+                            if available otherwise run on CPU.
+        """
+        import torch  # lgtm [py/repeated-import]
+        import yaml
+        from argparse import Namespace
+        from fairseq import checkpoint_utils, tasks, utils
+        from fairseq.data import encoders
+        import sentencepiece as spm
+
+        # Super initialization
+        super().__init__(
+            model=None,
+            clip_values=clip_values,
+            channels_first=None
```

Comment:
@YiwenShaoStephen I think this missing comma makes the CI cail:
```suggestion
            channels_first=None,
```